File: backend/src/game/actor.rs
use std::collections::VecDeque;

use axum::extract::ws::Message;
use tokio::sync::mpsc::UnboundedReceiver;
use tracing::info;

use crate::game::event::command::GameCommand;
use crate::game::event::EventManager;
use crate::game::config::GAME_TICK_MS;
use crate::game::map_data::ALL_MAP_DATA;
use crate::room::room::ClientDataForState;
use crate::state::AppState;
use crate::ws::message::{make_ws_msg_game_started, SendGameStartedPayload};

pub async fn run_game_actor(
    game_id: u32,
    players: Vec<ClientDataForState>,
    map_id: u32,
    mut rx: UnboundedReceiver<GameCommand>,
    state: &'static AppState,
) {
    let player_ids: Vec<u32> = players.iter().map(|p| p.id).collect();
    info!("game {} started with {} players", game_id, player_ids.len());

    let broadcast = |msg: Message, exclude_id: Option<u32>| {
        for &id in &player_ids {
            if Some(id) == exclude_id {
                continue;
            }
            state.send_to_client(id, msg.clone());
        }
    };

    let map_data = match ALL_MAP_DATA.get(&map_id) {
        Some(data) => data,
        None => {
            info!("map {} not found for game {}", map_id, game_id);
            broadcast(Message::Text("noMap".to_string()), None);
            return;
        }
    };

    let mut event_manager = EventManager::new(map_data, &players);

    let start_msg = Message::Text(
        make_ws_msg_game_started(SendGameStartedPayload {
            game_id,
            game_end_time: event_manager.game_end_time,
        })
        .into(),
    );
    broadcast(start_msg, None);

    let mut command_list: VecDeque<GameCommand> = VecDeque::new();
    let mut ticker =
        tokio::time::interval(std::time::Duration::from_millis(GAME_TICK_MS as u64));

    loop {
        tokio::select! {
            _ = ticker.tick() => {
                event_manager.consume_commands(&mut command_list, state);
                if event_manager.game_over {
                    break;
                }
            }
            Some(cmd) = rx.recv() => {
                command_list.push_back(cmd);
            }
            else => {
                break;
            }
        }
    }

    info!("game {} ended", game_id);
    state.game_sender_map.remove(&game_id);
}

File: backend2/game/actor.go
package game

import (
	"encoding/json"
	"log"
	"time"

	"bomberman/backend2/game/event"
	"bomberman/backend2/game/mapdata"
	"bomberman/backend2/state"
	"bomberman/backend2/types"
)

func makeWsMsg(msgType string, payload interface{}) []byte {
	b, _ := json.Marshal(map[string]interface{}{
		"type":    msgType,
		"payload": payload,
	})
	return b
}

func RunGameActor(
	gameID uint32,
	players []types.PlayerInfo,
	mapID uint32,
	rx <-chan state.GameCommand,
	appState *state.AppState,
) {
	playerIDs := make([]uint32, len(players))
	for i, p := range players {
		playerIDs[i] = p.ID
	}
	log.Printf("game %d started with %d players", gameID, len(players))

	broadcast := func(msg []byte) {
		for _, id := range playerIDs {
			appState.SendToClient(id, msg)
		}
	}

	mapData, ok := mapdata.AllMapData[mapID]
	if !ok {
		log.Printf("game %d: map %d not found", gameID, mapID)
		broadcast(makeWsMsg("error", "noMap"))
		return
	}

	em := event.NewEventManager(mapData, players)

	// Notify all players game started
	startMsg := makeWsMsg("gameStarted", map[string]interface{}{
		"gameId":      gameID,
		"gameEndTime": em.GameEndTime,
	})
	broadcast(startMsg)

	ticker := time.NewTicker(time.Duration(100) * time.Millisecond)
	defer ticker.Stop()

	var pending []state.GameCommand

	for {
		select {
		case <-ticker.C:
			em.ConsumeCommands(pending, appState)
			pending = pending[:0]
			if em.GameOver {
				goto done
			}
		case cmd, ok := <-rx:
			if !ok {
				goto done
			}
			pending = append(pending, cmd)
		}
	}

done:
	log.Printf("game %d ended", gameID)
	appState.GameSenderMap.Delete(gameID)
}
